homeassistant/components/sensor/__init__.py

```diff
@@ -127,3 +134,75 @@ def state_attributes(self) -> dict[str, Any] | None:
             return {ATTR_LAST_RESET: last_reset.isoformat()}
 
         return None
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the value reported by the sensor."""
+        return self._attr_native_value
+
+    @property
+    def native_unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement of the sensor, if any."""
+        return self._attr_native_unit_of_measurement
+
+    @property
+    def unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement of the entity, after unit conversion."""
+        if self._attr_unit_of_measurement is not None:
+            return self._attr_unit_of_measurement
+
+        native_unit_of_measurement = self.native_unit_of_measurement
+
+        if native_unit_of_measurement in (TEMP_CELSIUS, TEMP_FAHRENHEIT):
+            # Suppress AttributeError to handle tests testing directly on entity objects
+            with suppress(ValueError, AttributeError):
+                return self.hass.config.units.temperature_unit
+
+        return native_unit_of_measurement
+
+    @property
+    def state(self) -> Any:
```

Comment:
Should the state be marked as `final` now (or after the last conversion PR)?

Reply:
Yes, that's a good point. I think we can do it in this PR unless it breaks CI.